src/chain_parsers/visualsign-ethereum/src/abi_decoder.rs

```diff
@@ -0,0 +1,269 @@
+            };
+
+            // Simple decoding based on type
+            let formatted = decode_solidity_value(&input.ty, input_data, &mut offset);
+
+            let field = AnnotatedPayloadField {
+                signable_payload_field: SignablePayloadField::TextV2 {
+                    common: SignablePayloadFieldCommon {
+                        fallback_text: formatted.clone(),
+                        label: param_name,
+                    },
+                    text_v2: SignablePayloadFieldTextV2 { text: formatted },
+                },
+                static_annotation: None,
+                dynamic_annotation: None,
+            };
+            expanded_fields.push(field);
+        }
+
+        // Build function signature
+        let param_types: Vec<&str> = function.inputs.iter().map(|i| i.ty.as_str()).collect();
+        let signature = format!("{}({})", function.name, param_types.join(","));
+
+        let title = SignablePayloadFieldTextV2 {
+            text: function.name.clone(),
+        };
+
+        let subtitle = SignablePayloadFieldTextV2 {
+            text: signature.clone(),
+        };
+
+        Ok(SignablePayloadField::PreviewLayout {
+            common: SignablePayloadFieldCommon {
+                fallback_text: signature,
+                label: function.name.clone(),
+            },
+            preview_layout: SignablePayloadFieldPreviewLayout {
+                title: Some(title),
+                subtitle: Some(subtitle),
+                condensed: None,
+                expanded: if expanded_fields.is_empty() {
+                    None
+                } else {
+                    Some(SignablePayloadFieldListLayout {
+                        fields: expanded_fields,
+                    })
+                },
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const SIMPLE_ABI: &str = r#"[
+        {
+            "type": "function",
+            "name": "transfer",
+            "inputs": [
+                {"name": "to", "type": "address"},
+                {"name": "amount", "type": "uint256"}
+            ],
+            "outputs": [{"name": "", "type": "bool"}],
+            "stateMutability": "nonpayable"
+        },
+        {
+            "type": "function",
+            "name": "approve",
+            "inputs": [
+                {"name": "spender", "type": "address"},
+                {"name": "amount", "type": "uint256"}
+            ],
+            "outputs": [{"name": "", "type": "bool"}],
+            "stateMutability": "nonpayable"
+        }
+    ]"#;
+
+    #[test]
+    fn test_decoder_creation() {
+        let abi: JsonAbi = serde_json::from_str(SIMPLE_ABI).expect("Failed to parse ABI");
+        let decoder = AbiDecoder::new(Arc::new(abi));
+
+        // Should be able to look up functions
+        let selector = [0xa9, 0x05, 0x9c, 0xbb]; // transfer selector
+        assert!(decoder.find_function_by_selector(&selector).is_some());
```

Comment:
nit: test the returned value as well to ensure the returned value is as expected


```suggestion
let function = decoder.find_function_by_selector(&selector);
assert!(function.is_some());
assert_eq!(function.unwrap().selector(), selector); // compare the actual value and ensure it found the right function
```